iot_api/user_api/repository/TagRepository.py

```diff
@@ -0,0 +1,117 @@
+    """
+    Get a tag with the given tag_id and organization_id. If not exists raise an
+    exception.
+    """
+    query = db.session.query(Tag).filter(Tag.id==tag_id, Tag.organization_id==organization_id)
+    return query.first()
+
+def update(tag_id, name, color, organization_id):
+    """
+    Update the name and/or color of the tag with the given tag_id and
+    organization_id. If the tag does no exists, raise an exception.
+    """
+    query = db.session.query(Tag).filter(Tag.id==tag_id, Tag.organization_id==organization_id)
+    tag = query.first()
+    if name: tag.name = name
+    if color: tag.color = color
+    db.session.commit()
+
+def delete(tag_id, organization_id):
+    """ 
+    Delete the tag with the given tag_id and organization_id. If not found,
+    raise and exception.
+    """
+    query = db.session.query(Tag).filter(Tag.id==tag_id, Tag.organization_id==organization_id)
+    tag = query.first()
+    db.session.delete(tag)
+    db.session.commit()
+
+def is_from_organization(tag_id, organization_id):
+    """
+    Return a boolean indicating if the tag belongs to this organization.
+    """
+    return db.session.query(Tag.query.filter(
+        Tag.id == tag_id,
+        Tag.organization_id == organization_id
+    ).exists()).scalar()
+
+def tag_asset(tag_id, asset_id, asset_type, organization_id):
+    """
+    Tag the asset with the given asset_type ("device" or "gateway") and asset_id
+    (device_id or gateway_id) with the tag with tag_id and organization_id.
+    """
+    if not is_from_organization(tag_id, organization_id):
+        raise Exception("Trying to use a tag from other organization.")
+
+    if asset_type=="device":
+        if not Devices.is_from_organization(asset_id, organization_id):
+            raise Exception("Trying to tag a device from other organization")
+        asset_tag = DeviceTag(tag_id=tag_id, device_id=asset_id)
+        db.session.add(asset_tag)
+    elif asset_type=="gateway":
+        if not Devices.is_from_organization(asset_id, organization_id):
+            raise Exception("Trying to tag a gateway from other organization")
+        asset_tag = GatewayTag(tag_id=tag_id, gateway_id=asset_id)
+        db.session.add(asset_tag)
+    else:
+        raise Exception(f"Invalid asset_type: {asset_type}")
+    db.session.commit()
+
+def untag_asset(tag_id, asset_id, asset_type, organization_id):
+    """
+    Remove the tag with the tag_id and organization_id from the asset with the
+    given asset_type ("device" or "gateway") and asset_id (device_id or
+    gateway_id).
+    """
+    if not is_from_organization(tag_id, organization_id):
+        raise Exception("Trying to delete a tag from other organization.")
+
+    if asset_type=="device":
+        if not Devices.is_from_organization(asset_id, organization_id):
+            raise Exception("Trying to tag a device from other organization")
+        asset_tag = DeviceTag(tag_id=tag_id, device_id=asset_id)
```

Comment:
Could be possible that a user sends a tag within the organization but the asset id from another organization?

Reply:
Both checks are made: the tag and the asset must belongs to the organization. This part of the code was rearranged to be clearer.